### PR TITLE
Fix map.getProjection() returns explicity set globe and simplifying Transform.clone()

### PR DESCRIPTION
**Pull request by SnailBones**
As of the issue draft [here](https://github.com/orgs/mapbox/projects/677/views/10):

> When doing `map.setProjection('globe')`, `map.getProjection()` should always return `'globe'` independently of whether the map has transitioned to mercator.
> 
> This behavior would be consistent with the way other projections work.

The single-source of truth for most projections is `map._runtimeProjection` as described [here](https://github.com/mapbox/mapbox-gl-js/pull/11204). However, this is not the case with Globe view, which overrides this variable at zoom levels that switch to Mercator and keeps track of the original projection with the `this._transitionFromGlobe`. This PR checks this variable to alter the return of `map.getProjection()'

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [X] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Changed map.getProjection to return globe even when fully zoomed in.</changelog>`

### CODE REVIEW

File: src/ui/map.js [hunk 1]
```diff
@@ -1025,7 +1025,11 @@ class Map extends Camera {
      * const projection = map.getProjection();
      */
     getProjection() {
-        return this.transform.getProjection();
+        const proj = this.transform.getProjection();
+        if (this._transitionFromGlobe) {
+            proj.name = "globe";
+        }
```
Comment (SnailBones) [inline src/ui/map.js:1031]: @karimnaaji I address the issue here in a superficial (hacky?) way. An alternative would be to change the internal behavior of globe to be more consistent with other projections. (Preserve "globe" as `map._runTimeProjection` and get rid of `map._transitionFromGlobe`) What do you think is the best path forward here?

Comment (karimnaaji) [reply src/ui/map.js:1031]: >(Preserve "globe" as map._runTimeProjection and get rid of map._transitionFromGlobe) What do you think is the best path forward here?

Could you store the implicit and explicit projection in different members so that you would just return it here instead of overriding? `_runTimeProjection` would be renamed `_explicitProjection` as the naming doesn't reflect anything _runtime_, and you would always return the `_explicitProjection` here but use the `_implicitProjection` internally whenever it is set.

`_transitionFromGlobe` could eventually be removed as it may reduce to checks against whether the implicit projection is globe or mercator.

File: src/ui/map.js [hunk 2]
```diff
@@ -1043,20 +1047,43 @@ class Map extends Camera {
...
-        this._transitionFromGlobe = false;
-        this.style.updateProjection();
+        this.updateProjection(projection);
         return this;
     }
 
-    _setImplicitProjection(projection: ProjectionSpecification) {
-        this._runtimeProjection = projection;
-        this.style.updateProjection();
+    updateProjection(projection?: ProjectionSpecification) {
```
Comment (mourner) [inline src/ui/map.js:1059]: As far as I understand, `updateProjection` isn't supposed to be exposed in a public API — we have `setProjection` for this, so `_updateProjection` would be a more fitting name. But looking further, at this point there doesn't seem to be much value in keeping these separate — `setProjection` is now just a wrapper around `updateProjection` that additionally handles the first call without a style (`this._lazyInitEmptyStyle()` will be a noop on subsequent calls), and string alias. So we might as well just inline `updateProjection` into `setProjection` and use the latter internally across the codebase?

**Comment by mourner** [comment]
> * <changelog>Changed map.getProjection to return globe even when fully zoomed in.</changelog>

@SnailBones Nit: since globe view isn't formally released, any changes to the behavior shouldn't be a part of the changelog.

File: src/ui/map.js (hunk 2 above)
Comment (SnailBones) [reply src/ui/map.js:1059]: Good idea @Vlad!

File: src/ui/map.js [hunk 3]
```diff
@@ -1043,20 +1047,43 @@ class Map extends Camera {
...
+        if (projection) { this._runtimeProjection = projection; }
+        const prevProjectionState = this.transform.projection;
+        // newProjection is a user-friendly projection object or null if not changed
+        const newProjection = this.transform.setProjection(this._runtimeProjection || (this.style.stylesheet ? this.style.stylesheet.projection : undefined));
+        const projectionState = this.transform.projection;
+
+        this.style.tryDraping();
```
Comment (karimnaaji) [inline src/ui/map.js:1066]: nit: I would rename that function since `tryDraping` suggests the action of draping, so it almost sounds like some draping rendering will happen when called. What this function does is trying to enable terrain to later _leverage_ draping.

File: src/ui/map.js [hunk 4]
```diff
@@ -1043,20 +1047,43 @@ class Map extends Camera {
      * @see [Example: Display a web map using an alternate projection](https://docs.mapbox.com/mapbox-gl-js/example/map-projection/)
      * @see [Example: Use different map projections for web maps](https://docs.mapbox.com/mapbox-gl-js/example/projections/)
      */
-    setProjection(projection?: ?ProjectionSpecification | string) {
+    setProjection(projection: ProjectionSpecification | string = {name: "mercator"}) {
```
Comment (karimnaaji) [inline src/ui/map.js:1050]: Are we at risk of breaking the public API with that particular function signature change? The only requirement here is that we should still allow `map.setProjection(null | undefined)` as mentioned in the docs as a valid call.

File: src/ui/map.js (hunk 3 above)
Comment (SnailBones) [reply src/ui/map.js:1066]: Good catch @karimnaaji!

File: src/ui/map.js (hunk 4 above)
Comment (SnailBones) [reply src/ui/map.js:1050]: Good catch Karim! It works on undefined but no longer on `map.setProjection(null)`. I'll fix this and add a test.

Comment (karimnaaji) [reply src/ui/map.js:1050]: Thanks! Yes any unit tests coverage increase of APIs touching projections is a good call, the coverage is still fairly low so we need to be careful.

File: src/ui/map.js (hunk 2 above)
Comment (SnailBones) [reply src/ui/map.js]: I've combined `updateProjection` and `setProjection` but it's breaking `null` expression fallback behavior. (This is the reason for the current failing test and is described [here](https://github.com/mapbox/mapbox-gl-js/pull/11204). Calls from the API need to save the  result of `transform.setProjection` as a high priority "explicit" expression while other projection updates (globe transition or style spec changes) should not. IMO the simplest way to maintain this is to keep the two as separate functions.

File: src/style/style.js [hunk 5]
```diff
@@ -352,16 +351,22 @@ class Style extends Evented {
...
-    updateProjection() {
-        const prevProjection = this.map.transform.projection;
-        const projectionChanged = this.map.transform.setProjection(this.map._runtimeProjection || (this.stylesheet ? this.stylesheet.projection : undefined));
-        const projection = this.map.transform.projection;
+    _updateProjection() {
+        if (!this.map._explicitProjection) {
+            this.map._updateProjection();
+        } else { // Ensure that style is consistent with current projection on style load
+            this.enableDraping();
+        }
```
Comment (karimnaaji) [inline src/style/style.js:364]: Should this logic/flow be part of and consolidated in `map._updateProjection`?

File: src/style/style.js [hunk 6]
```diff
@@ -352,16 +351,22 @@ class Style extends Evented {
         } else {
             delete this.stylesheet.projection;
         }
-        this.updateProjection();
+        if (!this.map._explicitProjection) {
+            this.map._updateProjection(projection);
+        }
```
Comment (karimnaaji) [inline src/style/style.js:356]: Similar comment as below (line 360-364), I think this could be compressed for an early return if moved within `map._updateProjection` so that you always call `map._updateProjection` from here.

File: src/ui/map.js [hunk 7]
```diff
@@ -1045,18 +1056,51 @@ class Map extends Camera {
...
+            this.painter.clearBackgroundTiles();
+            for (const id in this.style._sourceCaches) {
+                this.style._sourceCaches[id].clearTiles();
+            }
+        }
+        this.style.dispatcher.broadcast('setProjection', this.transform.projectionOptions);
+        this.style.enableDraping();
+
+        this._update(true);
+        return this;
```
Comment (karimnaaji) [inline src/ui/map.js:1103]: nit: for brevity
```suggestion
        return this._update(true);
```

File: src/ui/map.js [hunk 8]
```diff
@@ -2892,7 +2923,15 @@ class Map extends Camera {
...
-        this._updateProjection();
+        if (this.getProjection().name === 'globe') {
+            if (this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX) {
+                if (this.transform.projection.name === 'globe') {
+                    this._updateProjection();
+                }
+            } else if (this.transform.projection.name === 'mercator') {
+                this._updateProjection();
+            }
+        }
```
Comment (karimnaaji) [inline src/ui/map.js:2934]: Any way this logic be contained and simplified within `_updateProjection` to be closer to context?

File: src/ui/map.js [hunk 9]
```diff
@@ -367,9 +365,14 @@ class Map extends Camera {
...
     _averageElevation: EasedVariable;
-    _runtimeProjection: ProjectionSpecification | void | null;
     _containerWidth: number;
-    _containerHeight: number
+    _containerHeight: number;
+
+    // `_explicitProjection represents projection as set with a call to map.setProjection()
+    // For the actual projection displayed, use `transform.projection`.
+    // (The two diverge above the transition zoom threshold in Globe view or when _explicitProjection === null
+    // a null _explicitProjection indicates means the map defaults to first the stylesheet projection if present, then Mercator)
```
Comment (karimnaaji) [inline src/ui/map.js:374]:
```suggestion
    // a null _explicitProjection indicates the map defaults to first the stylesheet projection if present, then Mercator)
```

File: src/ui/map.js [hunk 10]
```diff
@@ -1045,18 +1056,51 @@ class Map extends Camera {
...
+                this._explicitProjection = {name: 'globe', center: [0, 0]};
+                changeExplicit = false;
+            }
+        }
+
+        const newProjection = this.transform.setProjection(projection);
+        if (!newProjection) return; // Continue if projection has changed
+
+        if (changeExplicit) {
+            this._explicitProjection = newProjection;
```
Comment (karimnaaji) [inline src/ui/map.js:1088]: I find the flow of this function quite difficult to follow, is there any way we could simplify it?
- Expose all early return conditions at the beginning of the function, otherwise wrap these entire blocks that need to be excluded in if statements, so that we don't have the [`return`](https://github.com/mapbox/mapbox-gl-js/pull/11460/files#diff-f62d8d42ec7424781f2e987a61684959e1bc794c67f75cd33548bd4a62c73028R1085) in the middle of the function
- Consolidate the cases for which we write the end result of `this._explicitProjection` and `changeExplicit` into a single write to make it easier to follow. We write to `this._explicitProjection` in 3 different locations in these statements.

File: src/ui/map.js [hunk 11]
```diff
@@ -1023,8 +1028,14 @@ class Map extends Camera {
...
      * const projection = map.getProjection();
      */
-    getProjection() {
-        return this.transform.getProjection();
+    getProjection(): ProjectionSpecification {
+        if (this._explicitProjection) {
+            return this._explicitProjection;
+        }
+        if (this.style && this.style.stylesheet && this.style.stylesheet.projection) {
+            return this.style.stylesheet.projection;
```
Comment (SnailBones) [inline src/ui/map.js:1036]: In the case where we fallback to a stylesheet projection, the projection will be returned in the same format it appears in the stylesheet. i.e. center and parallels can be absent.

This allows us to avoid storing another projection value (`map._runtimeProjection` or `map._implicitProjection`).

The potential downside I see in this is that it could impose a challenge for users when attempting to check if two projections are the same.

One other option to avoid that would be adding a `transform.standardizeProjection(projectionSpec)` function and standardize the projection here before returning.

@karimnaaji @ansis I'd love to hear your thoughts on what the right approach is!

File: src/style/style.js (hunk 5 above)
Comment (SnailBones) [reply src/style/style.js:364]: I don't think so. This logic is called when there are changes in the style sheet. If the map has an explicit projection, it overrides the style spec projection, so no changes to the map are necessary.

This check is only necessary on style sheet updates.

File: src/ui/map.js (hunk 8 above)
Comment (SnailBones) [reply src/ui/map.js:2934]: Similar to my comment in `style.js`, this logic determines when a projection update is needed on render (when passing the zoom threshold). It couldn't go in `map._updateProjection()` because the check is irrelevant during projection changes for any other reason.

It could be wrapped in its own function (as it was previously), but this is the only place the function would be called.

File: src/ui/map.js [hunk 12]
```diff
@@ -2892,7 +2922,15 @@ class Map extends Camera {
...
-        this._updateProjection();
+        if (this.getProjection().name === 'globe') {
+            if (this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX) {
+                if (this._runtimeProjection && this._runtimeProjection.name === 'globe') {
+                    this._updateProjection();
+                }
+            } else if (!this._runtimeProjection || this._runtimeProjection.name === 'mercator') {
+                this._updateProjection();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (SnailBones) [reply src/ui/map.js]: These don't belong in `_updateProjection`, since zoom transitions have no bearing during calls to setProjection().

The check here ensures that projection updates when the zoom threshold is passed, while the logic inside of `updateProjection` determines the correct projection. It's important to separate these two checks because if the user zooms in before changing the projection to globe, the map should render as Mercator.

The logic here could be placed in its own function, but this is the only place where it would be called.

File: src/ui/map.js (hunk 11 above)
Comment (karimnaaji) [reply src/ui/map.js:1036]: >The potential downside I see in this is that it could impose a challenge for users when attempting to check if two projections are the same.
>
>One other option to avoid that would be adding a transform.standardizeProjection(projectionSpec) function and standardize the projection here before returning.

That's a good point, but probably good to follow up after this PR is finalized to make sure it doesn't grow too much out of scope. But the standardization step sounds good to me.

File: src/ui/map.js [hunk 13]
```diff
@@ -366,9 +366,11 @@ class Map extends Camera {
     _silenceAuthErrors: boolean;
     _averageElevationLastSampledAt: number;
     _averageElevation: EasedVariable;
-    // In high zoom levels on Globe, _runtimeProjection will be "Mercator" while _explicitProjection will stay "Mercator"
+    // In high zoom levels on Globe, _runtimeProjection will be "mercator" while _explicitProjection will stay "globe"
+    // a null _runtimeProjection indicates means the stylesheet projection will be used
```
[the remark this answers is not in the document]
Comment (SnailBones) [reply src/ui/map.js]: Good catch!

File: src/ui/map.js (hunk 8 above)
Comment (karimnaaji) [reply src/ui/map.js:2935]: Have you investigated whether it is possible to split `this._updateProjection` into separate functions accommodating for the case we intend to do from the callee?
- One for the stylesheet/user update
- One function for the transition switch for explicit projection

I'm a bit concerned about the complexity of these checks, because if we expand/inline what we're doing for this specific case, we end up going through a lot of redundancy:
```
if (this.getProjection().name === 'globe') {
  if (this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX) {
    if (this.transform.projection.name === 'globe') {
      ... from within _updateProjection()
      const prevProjection = this.getProjection();
      ...
      if (projection && projection.name === "globe") {
        projection = {name: (this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX ? 'mercator' : 'globe')};
      }
      ...
```

File: src/ui/map.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (karimnaaji) [reply src/ui/map.js]: Refer to my comment in https://github.com/mapbox/mapbox-gl-js/pull/11460#discussion_r802121401, maybe there's room for splitting things up to reduce the scope of `_updateProjection` for a more lean and specific case between what the stylesheet wants to do when updating the projection vs what is needed for a transition.

File: src/ui/map.js (hunk 11 above)
Comment (karimnaaji) [reply src/ui/map.js:1036]: Captured in https://github.com/mapbox/mapbox-gl-js/issues/11490 for follow up, marking this comment as resolved.

File: src/ui/map.js (hunk 10 above)
Comment (SnailBones) [reply src/ui/map.js]: Pushed your first suggestion and I'm looking into the second.

File: src/ui/map.js [hunk 14]
```diff
@@ -1045,18 +1056,52 @@ class Map extends Camera {
      */
     setProjection(projection?: ?ProjectionSpecification | string) {
         this._lazyInitEmptyStyle();
-        if (typeof projection === 'string') {
+        if (projection === undefined) { projection = null; } else if (typeof projection === 'string') {
```
Comment (karimnaaji) [inline src/ui/map.js:1059]: I don't have anything against one liner for a single `if` statement, but let's expand this one into separate 
statements.

Although, it's probably not really a hot debugging codepath here, it can help the debugging experience to be able to put breakpoints in separate lines based on the condition.
```suggestion
        if (projection === undefined) {
            projection = null;
        } else if (typeof projection === 'string') {
```

File: src/ui/map.js (hunk 8 above)
Comment (SnailBones) [reply src/ui/map.js:2929]: In my first attempt, the projection was always passed into the equivalent of `_updateProjection` and in this case we could instead call a similar function `_updateGlobeTransition` that could skip the redundant checks for globe and zoom.

But by doing so, we would lose the advantages to having all projection changes happen in one function. Most important of which is that this ensures that projection updates return consistent results from application state, regardless of the function that modified the state. This is helpful for avoiding buggy edge cases that can be difficult to think through or test for all possible combinations. For example, prior to these changes, updating the projection to globe by changing the style spec would always render globe instead of Mercator at high zooms.

I've made some changes to `_updateProjection` and replaced some of the if statements with nested ternary operators. Let me know if you think that helps.

Comment (karimnaaji) [reply src/ui/map.js:2929]: Yes that helps, your changes in _updateProjection are simpler to follow.

**Review by karimnaaji — COMMENTED** [review]
Based on our 1:1 convo the last things to look at before merging:
- Add a few comments to https://github.com/mapbox/mapbox-gl-js/pull/11460/files#diff-f62d8d42ec7424781f2e987a61684959e1bc794c67f75cd33548bd4a62c73028R2921-R2929
- Add render tests for heatmap/circle with mercator -> globe -> mercator switches covering the case described in https://github.com/mapbox/mapbox-gl-js/pull/11437#discussion_r793539259, as it is an easy code path to break and is something that could be affected by this PR

After that it's ready to go :green_circle: nice work on that PR @SnailBones !

**Review by karimnaaji — APPROVED** [review]
